packages/compiler/src/optimizer/optimize.ts

```diff
@@ -0,0 +1,115 @@
+} from "@kuma-ui/core/components/componentList";
+import {
+  JsxOpeningElement,
+  JsxSelfClosingElement,
+  Node,
+  SyntaxKind,
+} from "ts-morph";
+import { isStyledProp, isPseudoProps } from "@kuma-ui/system";
+
+/**
+ * Optimizes JSX elements by converting Kuma components to raw HTML elements
+ * during compilation, when possible.
+ *
+ * The transformation can only be applied when no runtime-affecting prefix
+ * utility props are present. This is because these would necessitate runtime
+ * processing and therefore an intermediary component.
+ *
+ * For example, a <Flex as='section'> component would be converted to a <section>
+ * based on the 'as' prop, or to the default HTML element for that component.
+ *
+ * @param {string} componentName - The name of the Kuma component.
+ * @param {JsxOpeningElement|JsxSelfClosingElement} jsxElement - The JSX element to optimize.
+ * @param {string|undefined} as - The HTML element to use as a replacement.
+ */
+export const optimize = (
+  componentName: (typeof componentList)[keyof typeof componentList],
+  jsxElement: JsxOpeningElement | JsxSelfClosingElement,
+  as?: string
+) => {
+  const isOptimizable = jsxElement.getAttributes().every((attrLike) => {
+    if (Node.isJsxSpreadAttribute(attrLike)) return false;
+    const attr = attrLike.asKindOrThrow(SyntaxKind.JsxAttribute);
+    if (hasDynamicProp(attr.getNameNode().getText())) return false;
+    return true;
+  });
+
+  if (!isOptimizable) return;
+
+  const rawHTMLTag = (() => {
+    const tag = defaultComponentTag[componentName];
+    if (!!as && typeof as === "string") {
+      return as.replace(/['"`]/g, "");
+    } else {
+      if (typeof tag === "string") return tag;
+      return "div";
+    }
+  })();
+
+  safeReplaceTagName(jsxElement, rawHTMLTag);
+};
+
+function hasDynamicProp(key: string): boolean {
+  return (
+    isStyledProp(key) || isPseudoProps(key) || key === "variant" || key === "as"
+  );
+}
+
+/**
+ * Safely attempts to perform a replacement operation on a JSX element.
+ * If any error occurs during the operation, it is caught silently, ensuring
+ * that the rest of the code execution is unaffected.
+ *
+ * Notably, this function addresses a known edge case where replacements
+ * fail when a parent component has an 'as' prop and a child component doesn't.
+ *
+ * @param {JsxOpeningElement|JsxSelfClosingElement} jsxElement - Target JSX element to be replaced
+ * @param {string} newTagName - HTML tag to replace with
+ */
+function safeReplaceTagName(
+  jsxElement: JsxOpeningElement | JsxSelfClosingElement,
+  newTagName: string
+): void {
+  const originalComponent = jsxElement.getTagNameNode().getText();
+  try {
+    if (Node.isJsxOpeningElement(jsxElement)) {
+      const jsxElementParent = jsxElement.getParentIfKind(
+        SyntaxKind.JsxElement
+      );
+      if (jsxElementParent) {
+        jsxElementParent
+          .getOpeningElement()
+          .getTagNameNode()
+          .replaceWithText(newTagName);
```

Comment:
This is fine:
```tsx
<Box as='section' >
   <Box as='div' />
</Box>
```

This also works fine:
```tsx
<Box>
   <Box as='div' />
</Box>
```

And this is fine too:
```tsx
<Box>
   <Box/>
</Box>
```

However, it somehow fails only when the parent component has an 'as' prop and the child component doesn't:
```tsx
<Box as='section'>
   <Box/>
</Box>
```

Reply:
The last pattern works in my environment 🤔 

output:
```tsx
<Box as="section" className={"kuma-1218924623"}>
  <div className={"kuma-1218924623"} />
</Box>
```